### DISCUSSION TITLE
Is it possible to test Avalonia under MSTest?

### DISCUSSION THREAD

**Discussion by IronEaterBeast** [AvaloniaUI/Avalonia#15050] [Q&A]
I saw the section about XUnit in the documentation, but MSTest seems to integrate better with VS. Is it possible to test the process of running a window with MSTest? I don't need to test the UI part, I want to test the running process. 

I tried, but the test didn't run successfully. I encountered an error when using the code 'MainWindow oneMainWindow = new MainWindow();' - 'System.InvalidOperationException: Unable to locate 'Avalonia.Platform.IWindowingPlatform'.'

**Comment by thevortexcloud** [comment]
>  but MSTest seems to integrate better with VS. 

It does not really as far as I am aware. They all show tests in the same test window in the exact same way. The old MSTest is also a pain to use since it ships with Visual Studio and nowhere else. The new version is at least a nuget package, but it has differences.

Anyway, Avalonia does have a headless backend for this sort of thing (which means it wont try to use an actual UI) but I have no idea how to use it.

https://docs.avaloniaui.net/docs/concepts/headless/

Is there a particular reason you need to open an actual Window? That's not something you would typically write tests for since it's fairly hard to unit test.

**Reply by IronEaterBeast** [reply]
I'm used to simply right-clicking on a method to generate tests using MSTest, and I find it to be a very simple method. Since I haven't used XUnit, I can't imagine how it could be simpler. So, this is just my guess - perhaps XUnit isn't any harder than MSTest. However, my project is on a tight schedule, and I don't have time to learn how to use XUnit now. If MSTest can solve the problem, I don't see the need to risk switching to XUnit at the moment.

I've already tested most of the methods that can be using MSTest, but there are still some processes that remain in avalonia windows. I have to find a way to test them, otherwise I'm worried they might become a source of trouble.

**Comment by maxkatz6** [comment]
It is possible to use headless testing with any testing framework. But you might need to write more code to support these.
In general, see this manual setup of headless session:
https://docs.avaloniaui.net/docs/concepts/headless/headless-custom

You need to call HeadlessUnitTestSession.StartNew once on common setup.
And then `session.Dispatch(() => { actual testing code })` from each your test

**Reply by IronEaterBeast** [reply]
I've learned how to use it, it's very simple.

**Reply by BrianKiarieMwaniki** [reply]
Please share how you archived to unit test Avalonia with MsTest. At the moment I keep on getting a `System.InvalidOperationException: Call from invalid thread`

**Reply by IronEaterBeast** [reply]
Just follow the guide https://docs.avaloniaui.net/docs/concepts/headless/headless-custom

**Reply by BrianKiarieMwaniki** [reply]
Thanks, I've managed to unit test my app. Like you said, it's very simple.🙂
